### ISSUE TITLE
Simultaneous button clicks possible because of mousedown and mouseup timing

### ISSUE THREAD

**Issue by 96LawDawg** [ArnoldSmith86/virtualtabletop#929]
Say there is a button in a room.  Player 1 clicks it and holds the mousedown.  Player 2 clicks it and completes mousedown and mouseup.  Then Player 1 releases mousedown.  The button executes twice.  This can potentially cause a game to malfunction or other problems.  @ArnoldSmith86 and I discovered this by examining a trace in a room where this happened.

I have simplified the problem and created a demo room.  https://virtualtabletop.io/2hyr.  Click on the button and hold it down on your desktop.  Click and release on a mobile device.  The button will move x + 100.  Then release your desktop and the button will move x + 100 again.  This does not happen unless you hold the mouse still.  Moving the mouse while mousedown does not cause the problem.

I don't know how this could possibly be fixed, but I would categorize this as a very serious bug.

**Comment by 96LawDawg** [comment]
Also, adding a random element does not change this behavior because the room doesn't see it as a traditional conflict.  Click on the second button in my demo room using the same technique.  On mouseup on the second device, you get a random number and x + 100 movement.  On mouseup on the firsst device, you get another random number and x + 100 movement.

**Comment by robartsd** [comment]
I'm not sure how this differs from two players clicking the button close to the same time except that the button has moved out from under the cursor of the second mouseup event?

I think you're expecting a conflicting delta to reset the room of the second mouseup client. This does not happen because the first mouseup's delta has been processed before the second mouseup triggers the routine. If it is important to prevent the routine triggering again, the game designer should set clickable:false to prevent the second click (removing later when the button should be clickable again). If clickable gets set to false by the routine triggered by the first mouseup, the second mouseup will not trigger the routine.

**Comment by 96LawDawg** [comment]
I posted hoping for a solution.  You provided one.  Thank you.  So not a bug?  Fine, but not expected behavior by a casual player or designer.

**Comment by ArnoldSmith86** [comment]
We should still think about if this warrants a deeper change. Like only doing `click` if the cursor is still on the widget.

**Comment by robartsd** [comment]
I thought everyone expected a routine to run twice if the button is pressed twice.

The unexpected behavior is that the second button press succeeds even though the button is no longer under the cursor when the mouse is released. It might be possible to modify mousehandling.js to check that the target of the mousedown event is still under the cursor before calling the widget's `click()` method.

**Comment by ArnoldSmith86** [comment]
(And still visible for the clicking player.)

Maybe even check newly added occlusion but I think that's overkill.

**Comment by 96LawDawg** [comment]
Nobody (except programmers) expect this behavior.

**Comment by 96LawDawg** [comment]
Sorry, hit submit before I was done...

The fix shouldn't be limited to just movement or visibility.  Regular buttons (like the deal button in my poker room) that don't disappear or move do this (now that I know what to look for).  Dealing 2 cards instead of 1 because of button timing is not cool.  And resetting a clickable:false condition is a pain in most rooms.  

Could or should a clickRoutine not execute until mouseup?  I know we need mousedown for dragging cards, drawing on canvas, etc.  But it isn't reasonable to expect that a clickRoutine will run until the mouse click is finished.  And allowing that messes with room.  So is checking for mouseup for clickRoutines possible?

**Comment by robartsd** [comment]
One thing I don't understand about the mousehandling.js code is why mouse events need to store the mousedown target in mouseTarget, but touch events do not.

**Comment by ArnoldSmith86** [comment]
That's just the way the events work. I don't know why but touch events give you the target of the touchdown whenever there is a new event for the same touch.

Mouse events don't work like that. I think they always give you the current target. So if you drag a widget underneath another, it would give you the upper widget while the cursor is above that widget. Something like that.

**Comment by robartsd** [comment]
> That's just the way the events work. I don't know why but touch events give you the target of the touchdown whenever there is a new event for the same touch.
> 
> Mouse events don't work like that. I think they always give you the current target. So if you drag a widget underneath another, it would give you the upper widget while the cursor is above that widget. Something like that.

I'm guessing this is mostly because touch events were designed later with much more nuanced behaviors expected.

This difference could make it even harder to fix this for touchscreens.

**Comment by ArnoldSmith86** [comment]
> The fix shouldn't be limited to just movement or visibility. Regular buttons (like the deal button in my poker room) that don't disappear or move do this (now that I know what to look for). Dealing 2 cards instead of 1 because of button timing is not cool. And resetting a clickable:false condition is a pain in most rooms.

At some point we can't solve this problem though. Because of network latency I'm pretty sure we can't be 100% sure in every case if the clicks started simultaneously or one after the other.

How is VTT supposed to know if two "simultaneous" clicks should be one or two clicks? If the button only deals cards to _your own_ hand, it should probably be two. But it might not. If it deals to all hands, it should probably only do it once. But it might not.

We could add something like a `clickCooldown` property that only allows the `clickRoutine` to be executed again after the given amount of milliseconds have passed. But it still needs to be set by the game designer because it's just a case by case problem.

**Comment by ArnoldSmith86** [comment]
Don't get me wrong, I think we should improve this. But it's not an "obvious" solution.

For example I like adding draw buttons to hands. I absolutely want them to be clicked simultaneously. Otherwise players think "Why don't I have any cards? I clicked the button!".

**Comment by ArnoldSmith86** [comment]
As far as I can tell, `clickable` is checked on mouseup. So you should be able to prevent double clicks by setting `clickable: false` and a random value somewhere in the room (for example calling `SHUFFLE`).

We could add a check that doesn't count a click if the widget is no longer under the cursor on mouseup but I feel like that only gives you a false sense of safety. It wouldn't help with the fact that the players could just click the button virtually at the same time.

That being said, I still think it's probably a good idea to add some kind of mechanism that will not allow a button to be clicked by multiple players "at the same time". I'm not sure how that would look like though. If you _don't_ modify the button when it's clicked, it would have to be something time based (which would probably have to be configurable - maybe something like `clickCooldown`?). Something time based can always be a problem if the network connection is subpar. If you _do_ modify the button in its `clickRoutine`, you might as well set it to `clickable: false`, I guess (plus something that causes a conflict on simultaneous clicks).

**Comment by 96LawDawg** [comment]
So that works.  My random thing was clicking a spinner.  The angle just keeps increasing in random intervals.  Of course, there are a finite number of angles and it is possible for the number to match, but unlikely.  Shuffle would probably be better, but storing a large random number on the button would also work.

BUT, it also seems that just setting clickable:false also works.  That probably shouldn't be the case based on what you all have said.  But somehow (unless I just am not doing it right), I can no prevent it by setting clickable:false on the button itself without anything random.  But unless I'm missing something, that should work.  P1 goes mousedown.  P2 goes mousedown and mouseup.  Routine runs; clickable is set to false.  P1 goes mouseup and the conflict is the difference in true/false of the clickable property of the button.  I guess the problem we are trying to fix is if P1 goes mouseup before P2's button has a chance to report to the server.  Then the randomness would prevent that.  Much harder to duplicate that scenario due to timing.

**Comment by ArnoldSmith86** [comment]
Yes, your scenario doesn't need a conflict. The `clickable: false` prevents a second click when the second client already knows about the first click. The random value prevents a second click when the second client _does not_ already know about the first click.

It should be fairly easy to reproduce though by clicking the button the same time on two devices.
